### PR TITLE
refactor: use list in `BodyAtom` (Datalog engine)

### PR DESCRIPTION
**Pull request by JonathanStarup**
Related to #3896.

### CODE REVIEW

**Comment by JonathanStarup** [comment]
@magnus-madsen This should be ready but the community had a timeout? I'm not sure what that is about.

**Comment by magnus-madsen** [comment]
> @magnus-madsen This should be ready but the community had a timeout? I'm not sure what that is about.

That's strange.

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 1]
```diff
@@ -84,6 +84,8 @@ object Lowering {
 
     lazy val Comparison: Symbol.EnumSym = Symbol.mkEnumSym("Comparison")
     lazy val Boxed: Symbol.EnumSym = Symbol.mkEnumSym("Boxed")
+
+    lazy val FlixList: Symbol.EnumSym = Symbol.mkEnumSym("List")
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:88]: Rename to `FList` (thats the convention we use elsewhere).

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 2]
```diff
@@ -119,6 +121,7 @@ object Lowering {
     lazy val Comparison: Type = Type.mkEnum(Enums.Comparison, Nil, SourceLocation.Unknown)
     lazy val Boxed: Type = Type.mkEnum(Enums.Boxed, Nil, SourceLocation.Unknown)
 
+    def flixList(t: Type): Type = Type.mkEnum(Enums.FlixList, List(t), SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:124]: Rename to `mkList`

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 3]
```diff
@@ -1248,6 +1251,21 @@ object Lowering {
     Expression.ArrayLit(exps, reg, tpe, eff, loc)
   }
 
+  /**
+    * Returns a list expression constructed from the given list of expressions `exps`.
+    */
+  private def mkList(exps: List[Expression], elmType: Type, loc: SourceLocation): Expression = {
+    val listType = Types.flixList(elmType)
+    val nil = mkTag(Enums.FlixList, "Nil", Expression.Unit(loc), listType, loc)
+    def consOf(elm: Expression, rest: Expression): Expression = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1260]: Introduce two helper functions (below): `mkNil` and `mkCons` and use them here.

File: main/src/library/Fixpoint/Ast/BodyPredicate.flix [hunk 4]
```diff
@@ -58,19 +59,20 @@ namespace Fixpoint/Ast {
...
             match body {
                 case BodyAtom(predSym, Relational, p, f, terms) =>
-                    "${polarityPrefix(p)}${fixityPrefix(f)}${predSym}(${commaSeparate(terms)})"
+                    "${polarityPrefix(p)}${fixityPrefix(f)}${predSym}(${terms |> List.map(toString) |> String.intercalate(", ")})"
                 case BodyAtom(predSym, Latticenal(_), p, f, terms) =>
-                    let len = Array.length(terms);
-                    let k = terms[0..len-1];
-                    let l = terms[len-1];
-                    "${polarityPrefix(p)}${fixityPrefix(f)}${predSym}(${commaSeparate(k)}; ${l})"
+                    let (k, lList) = List.splitAt(List.length(terms)-1, terms);
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Ast/BodyPredicate.flix:64]: I propose something like `keys` and `lattice` or `ks` and `l` or similar or `keyTerms` and `latticeTerm`?

File: main/src/library/Fixpoint/Ast/BodyPredicate.flix [hunk 5]
```diff
@@ -58,19 +59,20 @@ namespace Fixpoint/Ast {
...
+                        case Some(l) => "${polarityPrefix(p)}${fixityPrefix(f)}${predSym}(${k |> List.map(toString) |> String.intercalate(", ")}; ${l})"
+                    }
                 case Guard0(_) => "<clo>()"
                 case Guard1(_, v) => "<clo>(${v})"
                 case Guard2(_, v1, v2) => "<clo>(${v1}, ${v2})"
                 case Guard3(_, v1, v2, v3) => "<clo>(${v1}, ${v2}, ${v3})"
                 case Guard4(_, v1, v2, v3, v4) => "<clo>(${v1}, ${v2}, ${v3}, ${v4})"
                 case Guard5(_, v1, v2, v3, v4, v5) => "<clo>(${v1}, ${v2}, ${v3}, ${v4}, ${v5})"
                 case Loop1(_) => "<loop>"
-            } as & Pure
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Ast/BodyPredicate.flix:74]: Nice to see that it becomes pure 👍

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**